src/plugins/discover/public/application/main/data_fetching/fetch_documents.ts

```diff
@@ -67,7 +69,15 @@ export const fetchDocuments = (
     .pipe(
       filter((res) => !isRunningResponse(res)),
       map((res) => {
-        return buildDataTableRecordList(res.rawResponse.hits.hits as EsHitRecord[], dataView);
+        return buildDataTableRecordList(res.rawResponse.hits.hits as EsHitRecord[], dataView, {
+          processRecord: (record) => {
+            profilesCollector.collect({ record });
+            return record;
+          },
+        });
```

Comment:
I was experimenting with optimizing records flattening by using a "lazy" getter function and I think it can become in handy for document context resolution too.  We could add a "lazy" getter for this logic to the new `DataTableRecord` class  too. This way we would process records "on demand" and not all at once which should make things faster.

https://github.com/jughosta/kibana/pull/11/files#diff-eebdb51e43f4e4e7083c39bddb64144d3e5da051d56d2fa68efb2d0a87479dc4R75

Reply:
Good idea! Document context resolution is synchronous anyway, so a lazy getter should do the trick. I wouldn't expect the resolution process to be heavy in general, but it seems reasonable to avoid running it until necessary. I was able to implement it already, but it could certainly be improved if we switch to a class later: https://github.com/elastic/kibana/pull/183797/commits/d8351ca9b511be5338740760c1ac3d22a5b3c3e5#diff-8501e1b7df4eeb0b27b10309dd2ffbe366964c264334485dee5a1d404e51a583R104-R108.